From 3df12c8c4ac43dfc09316a0c660e7c186a3508ba Mon Sep 17 00:00:00 2001
From: Neon Chen <neon_chen@tw.shuttle.com>
Date: Wed, 12 Oct 2016 17:41:44 +0800
Subject: [PATCH 315/443] Fix Face Identify API: If TCIT group error, it still
 work.

If there're more than one group in TCIT.
Identify personid in all the groups.
---
 api/recognition.py | 34 +++++++++++++++-------------------
 1 file changed, 15 insertions(+), 19 deletions(-)

diff --git a/api/recognition.py b/api/recognition.py
index 5ce58e8..2b21fc2 100644
--- a/api/recognition.py
+++ b/api/recognition.py
@@ -107,6 +107,7 @@ def face_identify(request):
                     face_id = res_face_detect['faces'][0]['face_id']
 
                     # TCIT : identify person and face in group
+                    person_similarity = settings.CONFIDENCE[ip_device.FaceBioScore]
                     for group in TCIT_API('/face/query/group_list')['groups']:
                         res_group_identify = TCIT_API(
                             '/face/group/identify',
@@ -114,29 +115,24 @@ def face_identify(request):
                         )
 
                         if res_group_identify['state'] == TCIT_res_state["STATE_SUCCESSFUL"]:
-                            # sort persons and get max similarity
-                            person = sorted(
-                                res_group_identify['persons'],
-                                key = itemgetter('similarity'),
-                            )[len(res_group_identify['persons'])-1]
-                            if person['similarity'] >= settings.CONFIDENCE[ip_device.FaceBioScore]:
+                            person = res_group_identify['persons'][0]
+                            if person['similarity'] >= person_similarity:
+                                person_similarity = person['similarity']
                                 person_id = person['person_id']
-                            else:
-                                raise Exception("Similarity is too low")
                         else:
                             raise Exception('Request TCIT error')
-                        if person_id:
-                            fruser = Frusers.objects.get(PersonId = person_id, State = 'Enable')
-                            face_save(request, fruser.FRUserId)
+                    if person_id:
+                        fruser = Frusers.objects.get(PersonId = person_id, State = 'Enable')
+                        face_save(request, fruser.FRUserId)
 
-                            for member in fruser.frusergroupmember_set.filter(FRUserOId = fruser.OId):
-                                for device in member.GroupOId.frusergroupdevices_set.filter(GroupOId = member.GroupOId):
-                                    if device.DeviceOId == ip_device:
-                                        Fruserlogs.objects.create(FRUserOId = fruser.OId, FRUserId = fruser.FRUserId, FRUserName = fruser.Name, DeviceOId = ip_device.OId, DeviceName = ip_device.Name, LogContent = 'Face identify pass')
-                                        return JsonResponse(dict(ChainMap(result_pass, {'FRUserId' : fruser.FRUserId})))
-                        else:
-                            raise Exception('There is no person')
-                    raise Exception('There is no permission device')
+                        for member in fruser.frusergroupmember_set.filter(FRUserOId = fruser.OId):
+                            for device in member.GroupOId.frusergroupdevices_set.filter(GroupOId = member.GroupOId):
+                                if device.DeviceOId == ip_device:
+                                    Fruserlogs.objects.create(FRUserOId = fruser.OId, FRUserId = fruser.FRUserId, FRUserName = fruser.Name, DeviceOId = ip_device.OId, DeviceName = ip_device.Name, LogContent = 'Face identify pass')
+                                    return JsonResponse(dict(ChainMap(result_pass, {'FRUserId' : fruser.FRUserId})))
+                        raise Exception("Permission denied")
+                    else:
+                        raise Exception("Similarity is too low")
                 else:
                     raise Exception('Request TCIT error')
             else:
-- 
1.9.1

